src/passes/Heap2Local.cpp

```diff
@@ -265,6 +265,31 @@ struct Heap2LocalOptimizer {
       walk(func->body);
     }
 
+    // Rewrite the code in visit* methods. The general approach taken is to
+    // replace the allocation with a null reference (which may require changing
+    // types in some places, like making a block return value nullable), and to
+    // remove all uses of it as much as possible, using the information we have
+    // (for example, when our allocation reaches a RefAsNonNull we can simply
+    // remove that operation as we know it would not throw). Some things are
+    // left to other passes, like getting rid of dropped code without side
+    // effects.
+
+    void visitBlock(Block* curr) {
+      if (!reached.count(curr)) {
+        return;
+      }
+
+      // Our allocation passes through this block. We must turn its type into a
+      // nullable one, because we will remove things like RefAsNonNull of it,
+      // which means we may no longer have a non-nullable value as out input,
+      // and we could fail to validate. It is safe to make this change in terms
+      // of our parent, since we know very specifically that only safe things
+      // will end up using our value, like a StructGet or a Drop, which do not
+      // care about non-nullability.
```

Comment:
This is going to be slightly tricky to scale to multivalue tuples. Do we have tests that demonstrate what happens when values flow through tuples today?

Reply:
Currently flowing into a tuple will stop this optimization. We must explicitly handle everything we flow into, and anything not declared blocks (is considered to maybe escape).

Reply:
Yes, I was thinking that it would be useful to have tests for that so it is easy to see the behavior difference if we ever improve the pass to handle tuples in the future.